plutus-contract/src/Plutus/Trace/Emulator.hs

```diff
@@ -138,17 +145,31 @@ data PrintEffect r where
   PrintLn :: String -> PrintEffect ()
 makeEffect ''PrintEffect
 
-type EmulatorTrace =
-        Eff
-            '[ StartContract
-            , RunContract
-            , Assert
-            , Waiting
-            , EmulatorControl
-            , EmulatedWalletAPI
-            , LogMsg String
-            , Error EmulatorRuntimeError
-            ]
+data QueryChainState r where
+  GetChainState :: QueryChainState ChainState.ChainState
+  GetParams     :: QueryChainState Params
```

Comment:
It's true that I can get rid of `getChainState` but as far as I can tell the `getParams` effect is only available in the `Contract` monad. Is there a more general thing somewhere that I'm not aware of?

Reply:
You probably need to rebase first (it may be an awful rebase, sorry). And then this one should be accessible: https://github.com/input-output-hk/plutus-apps/blob/main/cardano-node-emulator/src/Cardano/Node/Emulator/Chain.hs#L97

Reply:
So `ChainEffect` is used elsewhere but isn't included in `EmulatorEffects` for some reason.

Reply:
A compromise is that we keep the `QueryParams` effects the way it is. The implementation overhead is minimal after all. Or alternatively we refactor this whole thing - but I'm not sure it would be good to include `ChainEffects` in the `EmulatorEffects` willy nilly as this causes some soundness issues for the testing model.

Reply:
Another option that I like more is to add `GetParams` also to `EmulatorControl` (as this is where the `ChainState` effect currently lives). That would reduce the number of effect types without putting `ChainEffect` in `EmulatorEffects`.

Reply:
Yes, I was thinking about it, it would be better than introducing an new type of effects for this specific purpose indeed.